src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py

```diff
@@ -0,0 +1,79 @@
+from core_data_modules.logging import Logger
+from engagement_database.data_models import MessageStatuses
+
+from src.engagement_db_to_rapid_pro.configuration import WriteModes
+
+log = Logger(__name__)
+
+# Value to write to a Rapid Pro contact field if we're only indicating the presence of an answer
+_PRESENCE_VALUE = "#ENGAGEMENT-DATABASE-HAS-RESPONSE"
+
+
+def sync_engagement_db_to_rapid_pro(engagement_db, rapid_pro, uuid_table, sync_config):
+    """
+    Synchronises an engagement database to Rapid Pro.
+
+    :param engagement_db: Engagement database to sync to.
+    :type engagement_db: engagement_database.EngagementDatabase
+    :param rapid_pro: Rapid Pro client to sync from.
+    :type rapid_pro: rapid_pro_tools.rapid_pro_client.RapidProClient
+    :param uuid_table: UUID table to use to de-identify contact urns.
+    :type uuid_table: id_infrastructure.firestore_uuid_table.FirestoreUuidTable
+    :param sync_config: Configuration for the sync.
+    :type sync_config: src.engagement_db_to_rapid_pro.configuration.EngagementDBToRapidProConfiguration
+    """
+    # Get all messages to sync.
+    # Only sync live messages. Anything else means we should signal Rapid Pro to request the data again if it can.
+    # TODO: This could get expensive as the number of datasets increases, in which case we can optimise by adding
+    #       an array-contains filter on the dataset field, so we only need to download messages in the datasets we
+    #       want to sync.
+    messages = engagement_db.get_messages(filter=lambda q: q.where("status", "in", [MessageStatuses.LIVE]))
+
+    # Organise messages by participant then by engagement db dataset.
+    participants = dict()  # of participant_uuid -> (dict of dataset -> list of Message)
+    for msg in messages:
+        if msg.participant_uuid not in participants:
+            participants[msg.participant_uuid] = dict()
+        participant = participants[msg.participant_uuid]
+
+        if msg.dataset not in participant:
+            participant[msg.dataset] = []
+        dataset = participant[msg.dataset]
+
+        dataset.append(msg)
+
+    # Make sure all the contact fields exist in the Rapid Pro workspace.
+    existing_contact_fields = [f.key for f in rapid_pro.get_fields()]
+    contact_fields_to_sync = [dataset_config.rapid_pro_contact_field for dataset_config in sync_config.normal_datasets]
+    for contact_field in contact_fields_to_sync:
```

Comment:
The ability to switch between live and dry runs would be great for this.

Reply:
Agreed, I'll add dry runs to the task list to be completed later